First pass grouping audio by app

XScorpion2 opened the pull request:
## Issues
 - Resolves #69 & #55 

## Description
Very roughly added in grouping audio sessions together by application. This allows you to control all audio sessions for a single application very similarly to how [EarTrumpet ](https://github.com/File-New-Project/EarTrumpet) handles it as well. In fact that is where the idea came from. To do this, this requires accessing the file name of the process main module, which can only be done reliably using a 64 bit process as 32 bit processes cannot access this information for 64 bit processes. The way this is implemented is by adding a AudioSessionGroup with similar API to AudioSessionWrapper, and then feeding all wrappers that came from the same file into the same session group. Since the APIs for Wrapper and Group are nearly identical, this would allow for a configuration option to enable or disable this grouping in the future if one so desires. Though tbh, after using the grouping by application of EarTrumpet, I never want to go back to ungrouped. In addition to this change, system sounds are now handled at the wrapper level, with a default display name of "System Sounds" so they can be controlled correctly.

## Types of changes
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] Requested changes are in a branch
- [x] Compiled and tested requested changes on target hardware (PC, device)
- [ ] Updated the documentation, if necessary
- [ ] Updated the LICENSES file, if necessary
- [x] Reviewed the [guidelines for contributing](../CONTRIBUTING.md) to this repository


## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

XScorpion2 committed:
First pass grouping audio by app

Updated application workflows to x64

Fallback for when you really can't access main module, also threading fixes

Found a solution that works smoother and should allow 32bit grabbing to 64bit names

Reverted change to 64 bit

t3knomanzer commented:
Thanks so much for this!
I'm finishing some work and I'll take a look asap, promised.

XScorpion2 commented:
Found an alternative solution to the 32 bit to 64 bit filename access problem. This also solved the security problem I was running into with few games I was testing.

t3knomanzer commented:
I just finished the feature I was working on and merged into master. 

Can't wait to take a look at this tomorrow! 
In the meantime, whenever you are ready, I'd recommend that you rebase on to master. You shouldn't have any conflicts, all my changes were on the firmware.

XScorpion2 commented:
Rebased & squashed

t3knomanzer commented:
> Found an alternative solution to the 32 bit to 64 bit filename access problem. This also solved the security problem I was running into with few games I was testing.

The solution is set to AnyCPU but "Favor 32-bit" is disabled. Therefore it runs as a 64-bit process on 64-bit systems and 32-bit on 32-bit systems.  
From 64 you can get info about 32 and 64 bit processes and from 32 you have no problem.

Whas there any other issue? What was your alternative?

XScorpion2 commented:
The solution was to use a native call into a kernal32.dll function to get the executive file name instead. This is done in AudioExtensions.cs

Besides the 32bit process accessing 64bit process file name issue, I was also seeing failures getting the file name for a few games that I assume are due to cheat protections in place. The new approach in AudioExtensions.cs also fixed that one.

t3knomanzer commented on Desktop/Application/MaxMix/Services/Audio/AudioSessionService.cs:
```diff
@@ -106,15 +107,15 @@ public void SetEndpointVolume(int volume, bool isMuted)
+        /// <param name="appID">The App Id of the target session.</param>
         /// <param name="volume">The desired volume.</param>
-        public void SetSessionVolume(int pid, int volume, bool isMuted)
+        public void SetSessionVolume(int appID, int volume, bool isMuted)
```
Rename method to SetSessionGroupVolume.

t3knomanzer commented on Desktop/Application/MaxMix/Services/Audio/AudioSessionService.cs:
```diff
@@ -140,31 +141,43 @@ private void InitializeSessionWrappers(AudioSessionManager2 sessionManager)
-                    var wrapper = RegisterSession(session);
-                    RaiseSessionCreated(wrapper.ProcessID, wrapper.DisplayName, wrapper.Volume, wrapper.IsMuted);
-                }
+                    RegisterSession(session);
```
We are skipping the system sounds session intentionally to reduce clutter. I would suggest not bringing it back unless users ask for it.
I think the philosophy should be to keep it simple and only add complexity as needed or requested.

XScorpion2 replied:
Hmm, I would argue that adding code to skip system sounds is intentionally adding code clutter to perfectly honest. As it's just a generic audio session control, all be it one spawned by windows itself. I can add the check back in if you feel it is necessary. Thoughts?

t3knomanzer commented on Desktop/Application/MaxMix/Services/Audio/AudioSessionService.cs:
```diff
@@ -140,31 +141,43 @@ private void InitializeSessionWrappers(AudioSessionManager2 sessionManager)
+                group.AddWrapper(wrapper);
+
+                _groups.Add(wrapper.AppID, group);
+                RaiseSessionCreated(wrapper.AppID, wrapper.DisplayName, wrapper.Volume, wrapper.IsMuted);
```
You should probably use the group properties instead of the wrapper properties as event arguments for consistency. Otherwise if 2 sessions that belong to the same parent process have different names, the event will be raised with different arguments each time which is confusing.

How are you dealing with the fact that multiple sessions for the same app may different volume levels when added? 

Shouldn't the wrapper volume levels be set to the group volume level?

t3knomanzer replied:
Rename RaiseSessionCreated to RaiseSessionGroupCreated. Since we are now dealing with session groups and not individual sessions, we don't really need to notify when a new session has been created but when a new group is.

t3knomanzer commented:
> The solution was to use a native call into a kernal32.dll function to get the executive file name instead. This is done in AudioExtensions.cs
> 
> 
> 
> Besides the 32bit process accessing 64bit process file name issue, I was also seeing failures getting the file name for a few games that I assume are due to cheat protections in place. The new approach in AudioExtensions.cs also fixed that one. 

Awesome! Thank you!
I have the review in progress, should be done soon. Thanks again, I love the idea.

t3knomanzer commented on Desktop/Application/MaxMix/Services/Audio/AudioSessionGroup.cs:
```diff
@@ -0,0 +1,112 @@
+        /// <summary>
+        /// The process ID of this session.
+        /// </summary>
+        private int ProcessID => _wrappers.First().Value.ProcessID;
```
Is this necessary?  
I don't thing the group should have a the concept of process id at all, that's something that just belongs in the session wrapper.

XScorpion2 replied:
It can be removed, I didn't want to change the main api entry points much when adding a group concept.

t3knomanzer commented on Desktop/Application/MaxMix/Services/Audio/AudioSessionGroup.cs:
```diff
@@ -0,0 +1,112 @@
+        /// <summary>
+        /// ???
+        /// </summary>
+        public bool IsSingleProcessSession => _wrappers.First().Value.IsSingleProcessSession;
```
Same here. I don't think this property is relevant to the group class and the value of the first wrapper may not be the same as the rest.

XScorpion2 replied:
It can be removed, I didn't want to change the main api entry points much when adding a group concept.

t3knomanzer commented on Desktop/Application/MaxMix/Services/Audio/AudioSessionGroup.cs:
```diff
@@ -0,0 +1,112 @@
+        /// <summary>
+        /// A reference to the process that created this session.
+        /// </summary>
+        public Process Process => _wrappers.First().Value.Process;
```
Same as above. I don't see the value of returning the process of the first session as the process of the group.

XScorpion2 replied:
A group doesn't have a process, as a group is a collection of multiple processes that were spawned with the same executable.

t3knomanzer commented on Desktop/Application/MaxMix/Services/Audio/AudioSessionGroup.cs:
```diff
@@ -0,0 +1,112 @@
+        /// <summary>
+        /// The display name of the process that created this session.
+        /// </summary>
+        public string DisplayName => _wrappers.First().Value.DisplayName;
```
Is there any way we can get the name of the application as it is displayed in the volume mixer here?  
If there are multiple sessions in the group, each one may have a different name. Returning the name of the first one seems confusing.

XScorpion2 replied:
That name is the MainWindowTitle name of the process, I can update the display name to that if desired. Though do keep in mind that some processes, such as audio players, change this title frequently with each new song, and sometimes get very verbose.

t3knomanzer commented on Desktop/Application/MaxMix/Services/Audio/AudioSessionGroup.cs:
```diff
@@ -0,0 +1,112 @@
+        /// </summary>
+        public int Volume
+        {
+            get => _wrappers.First().Value.Volume;
```
Why not have a private volume field in this class?   
I think the group should subscribe to the volume and name changed events of the wrappers, update it's fields and raise new events when changes occur.

t3knomanzer commented on Desktop/Application/MaxMix/Services/Audio/AudioSessionGroup.cs:
```diff
@@ -0,0 +1,112 @@
+        #endregion
+
+        #region Public Methods
+        public void AddWrapper(AudioSessionWrapper wrapper)
```
I would here subscribe to the wrapper events.

t3knomanzer commented on Desktop/Application/MaxMix/Services/Audio/AudioSessionGroup.cs:
```diff
@@ -0,0 +1,112 @@
+            _wrappers.Add(wrapper.ProcessID, wrapper);
+        }
+
+        public bool RemoveWrapper(AudioSessionWrapper wrapper)
```
Unsubscribe from wrapper events

t3knomanzer commented on Desktop/Application/MaxMix/Services/Audio/AudioSessionService.cs:
```diff
@@ -140,31 +141,43 @@ private void InitializeSessionWrappers(AudioSessionManager2 sessionManager)
             var wrapper = new AudioSessionWrapper(session);
-            _wrappers[wrapper.ProcessID] = wrapper;
-
             wrapper.SimpleVolumeChanged += OnSimpleVolumeChanged;
```
I would make groups responsible for subscribing and managing wrapper events and the service responsible for dealing with group events.

t3knomanzer commented on Desktop/Application/MaxMix/Services/Audio/AudioSessionService.cs:
```diff
@@ -140,31 +141,43 @@ private void InitializeSessionWrappers(AudioSessionManager2 sessionManager)
+            {
+                _groups.Remove(group.AppID);
+                group.Dispose();
+                RaiseSessionRemoved(wrapper.AppID);
```
Use the group properties and not the wrapper. Rename method to RaiseSessionGroupRemoved.

t3knomanzer commented on Desktop/Application/MaxMix/Services/Audio/AudioSessionService.cs:
```diff
@@ -194,21 +207,19 @@ private void OnEndpointVolumeChanged(object sender, AudioEndpointVolumeCallbackE
             UnregisterSession(wrapper);
         }
 
         private void OnSimpleVolumeChanged(object sender, AudioSessionSimpleVolumeChangedEventArgs e)
```
This should now be part of the AudioSessionGroup class.

t3knomanzer commented on Desktop/Application/MaxMix/Services/Audio/IAudioSessionService.cs:
```diff
@@ -12,7 +12,7 @@ internal interface IAudioSessionService
         void Stop();
         void SetEndpointVolume(int volume, bool isMuted);
-        void SetSessionVolume(int pid, int volume, bool isMuted);
+        void SetSessionVolume(int appID, int volume, bool isMuted);
```
Rename to SetSessionGroupVolume

t3knomanzer reviewed (changes requested):
Hey,
First of all, thanks a lot for the contribution!
The main issue that I see is with the separation of concerns.  
With the introduction of the AudioSessionGroup class, the responsibility of dealing with AudioSessionWrappers should belong it.   
The only interaction that AudioSessionService should have with AudioSessionWrappers is adding them to the right AudioSessionGroup when they are created. Other than that, it should just deal with groups.

Let me know what you think.  
Thanks a lot again, I'm so excited to see this!

XScorpion2 commented:
So I was responding to each comment one by one, but there is a bit of a pattern I think we should step back and discuss the approach to take. For starters, A group has no process, it's a collection of multiple processes for the same executable. This means that each process could potentially have it's own separate volume control. An example of this is Discord, you run it once, and it has 2 sound sessions, one for chat, one for notifications. And both could have separate volumes in the Windows Mixer.

Now since a group is no more than a collection of similar sessions, it didn't make sense for the group to handle new sessions as it would still need to be figured out if that new session required a new group or not. Thus this event remained on the Service level. For remove, I originally had groups handling removes, but found leaving it at the service level made more sense from a data oriented programming perspective as it allowed for more flexibility in future refactors. One such example is that a future change would be allowing optional grouping per similar sessions, all separate sessions, or some mix in between. By keeping the API of the group and sessions identical, we could extract that into an interface that is on both, and then at the Service level do all the logic necessary for deciding if the new session should be grouped or not.

So given that concept and flexibility of the data oriented layout, I left the new sessions, remove session, etc all where they were and in the group just decide how we want to work with multiple sessions. The generic "first" approach works, but maybe something more purpose driven is necessary. Thoughts?

t3knomanzer commented:
> So I was responding to each comment one by one, but there is a bit of a pattern I think we should step back and discuss the approach to take. For starters, A group has no process, it's a collection of multiple processes for the same executable. This means that each process could potentially have it's own separate volume control. An example of this is Discord, you run it once, and it has 2 sound sessions, one for chat, one for notifications. And both could have separate volumes in the Windows Mixer.

Yep, we are on the same page there :)

> Now since a group is no more than a collection of similar sessions, it didn't make sense for the group to handle new sessions as it would still need to be figured out if that new session required a new group or not. Thus this event remained on the Service level. 

Yeah that's exactly what I was referring to when I said:
*"The only interaction that AudioSessionService should have with AudioSessionWrappers is adding them to the right AudioSessionGroup when they are created"*

>For remove, I originally had groups handling removes, but found leaving it at the service level made more sense from a data oriented programming perspective as it allowed for more flexibility in future refactors. One such example is that a future change would be allowing optional grouping per similar sessions, all separate sessions, or some mix in between. By keeping the API of the group and sessions identical, we could extract that into an interface that is on both, and then at the Service level do all the logic necessary for deciding if the new session should be grouped or not.
> 
> So given that concept and flexibility of the data oriented layout, I left the new sessions, remove session, etc all where they were and in the group just decide how we want to work with multiple sessions. The generic "first" approach works, but maybe something more purpose driven is necessary. Thoughts?

After you explained the reasoning behind it makes a lot of sense why you made those decisions.  Sorry I should probably have asked before getting into the details of the code.
However I'd rather just keep it simple and do one thing at a time.   
Grouping is great, reduces clutter in the device interface and streamlines the user experience. It's in line with the philosophy of the project so let's just commit to it.
After it gets released, if users request being able to tweak sessions individually or having the option to configure it, then we can revisit it.

Does that sound good to you?

XScorpion2 commented:
👍 Sounds good, just wanted to make sure things were clear before I dove back in.

XScorpion2 committed:
Big audio session refactor

XScorpion2 commented:
@t3knomanzer Ok done with a big refactor. In light of the desire to keep things simple, I extracted nearly all the complexity of managing volumes for devices, groups, and sessions out into an IAudioSession interface. The service is now responsible for just doing the initial creation of one or more audio devices via AudioDevice class (IAudioSession interface, yes, we can now support any number of audio devices seamlessly), then firing off the device's InitializeSessions function. The rest is handled via the callbacks at the right level. I also added a setting to enable / disable showing of the system sounds for devices.

XScorpion2 committed:
Fixing minor display name issue

XScorpion2 committed:
Circular volume trigger fix

XScorpion2 committed:
Fixing whitespace in files

XScorpion2 commented:
Next feature I want to tackle after this is scroll the app name on the oled display

XScorpion2 committed:
Added appropriate value rounding to Volume return

t3knomanzer commented:
Hey, I'm getting a crash in AudioExtensions/GetMainModuleFileName as soon as I run it.
I've tried both running it in x86, x64.

`System.ComponentModel.Win32Exception
  HResult=0x80004005
  Message=Access is denied
  Source=System
  StackTrace:
   at System.Diagnostics.ProcessManager.OpenProcess(Int32 processId, Int32 access, Boolean throwIfExited)
   at System.Diagnostics.Process.GetProcessHandle(Int32 access, Boolean throwIfExited)
   at System.Diagnostics.Process.OpenProcessHandle(Int32 access)
   at System.Diagnostics.Process.get_Handle()
   at MaxMix.Services.Audio.AudioExtensions.GetMainModuleFileName(Process process, Int32 buffer) in D:\rhenares\Projects\RubenHenares\Hardware\MaxMix\maxmix-software\Desktop\Application\MaxMix\Services\Audio\AudioExtensions.cs:line 17
   at MaxMix.Services.Audio.AudioSessionService.OnSessionCreated(IAudioSession session) in D:\rhenares\Projects\RubenHenares\Hardware\MaxMix\maxmix-software\Desktop\Application\MaxMix\Services\Audio\AudioSessionService.cs:line 164
   at MaxMix.Services.Audio.AudioDevice.InitializeSessions() in D:\rhenares\Projects\RubenHenares\Hardware\MaxMix\maxmix-software\Desktop\Application\MaxMix\Services\Audio\AudioDevice.cs:line 132
   at MaxMix.Services.Audio.AudioSessionService.Initialize(Object stateInfo) in D:\rhenares\Projects\RubenHenares\Hardware\MaxMix\maxmix-software\Desktop\Application\MaxMix\Services\Audio\AudioSessionService.cs:line 119
   at System.Threading.QueueUserWorkItemCallback.WaitCallback_Context(Object state)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.QueueUserWorkItemCallback.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback()

  This exception was originally thrown at this call stack:
    [External Code]
    MaxMix.Services.Audio.AudioExtensions.GetMainModuleFileName(System.Diagnostics.Process, int) in AudioExtensions.cs
    MaxMix.Services.Audio.AudioSessionService.OnSessionCreated(MaxMix.Services.Audio.IAudioSession) in AudioSessionService.cs
    MaxMix.Services.Audio.AudioDevice.InitializeSessions() in AudioDevice.cs
    MaxMix.Services.Audio.AudioSessionService.Initialize(object) in AudioSessionService.cs
    [External Code]`

Any chance you could take a look?

XScorpion2 commented:
Ya, let me dig into that some more.

XScorpion2 commented:
By chance can you send me some more details about the process being accessed? Is it a 64 bit process, game, etc. Are you running maxmix in 32 bit mode, what is the display name of the audio session?

XScorpion2 committed:
Using less access privileges to get file name.

XScorpion2 commented:
Also just pushed a change to use a more limited access call to get the file name. Based on reading the C# source, it was getting the Handle using full access which is overkill.

t3knomanzer commented:
Sorry. It’s with both, Spotify and VLC, not sure if they are x64 or x86, I can check in a few when I get back home. 
I’ve tried running Maxmix in both x64 and x86. 
I can send you a crash dump if it helps.

XScorpion2 commented:
I've got VLC, but not Spotify. Will test my access modifier change with both of those shortly.

t3knomanzer commented:
> I've got VLC, but not Spotify. Will test my access modifier change with both of those shortly.

It works fine now, thank you!

t3knomanzer commented:
Thanks a lot for the changes, I really like the approach you've taken this time, it's lot cleaner now IMO.
I feel like grouping should be handled by the AudioDevice, it would make it even more clear.
What do you think?

XScorpion2 committed:
Minor cleanup

t3knomanzer reviewed (commented):
Thanks a lot for the changes, I really like the approach you've taken this time, it's lot cleaner now IMO.
I feel like grouping should be handled by the AudioDevice, it would make it even more clear.
What do you think?

XScorpion2 commented:
I'm on the fence. On one hand, sessions come from an AudioDevice, so it seems like the correct approach to have them handled there as it fits the object oriented design. However, the flip side is that this then puts a requirement on any future settings / options to push them down to the AudioDevice level in order for them to handle things like Grouping vs Non-Grouping options, see System Sound visibility for a partial example. However with the current setup, settings can just be pushed into the system and managed there instead of having to go into what is basically data which fits a data oriented design.

I obviously leaned towards the data oriented design with both approaches as it really improves how quickly things can be changed or adapted. For example, there could be a mode where multiple AudioDevices are managed in a group, and application volumes are grouped and kept in sync across audio devices. Something like this really speaks to the flexibility of the current approach, where pushing grouping and session management into the AudioDevice would make this much harder to implement.

I'm find with which ever approach you would prefer on this project and just providing my opinion between the two.

t3knomanzer commented:
I usually lean towards the object oriented approach since it's easier to understand responsibilities based on hierarchy but I also totally see your point and makes complete sense.  

For reference, the philosophy with this project is to keep it simple and only add complexity as needed. This applies to everything, from the code to the hardware and user interface.  
Balancing flexibility and complexity is key. Adding an option to enable/disable something adds complexity to the code and the user interface, hence deciding if it's really necessary, if the majority of users will need it it's what matters.  

That's the reasoning behind all this back and forth with grouping. On one hand you wan't to stay flexible to account for the "what ifs" but on the other you wonder if it's worth it.  
In that situation, the way I approach it is, let's keep it simple for now and add complexity as needed.  This is an iterative process, things can change over time.
Following that reasoning, let's push groups to the device level now, keep it simple and clear, favor an object oriented approach. Later when/if we want to add more complex functionality, we can revisit and change as needed.

Does that make sense? I hope It doesn't come across as me being prescriptive. On the contrary, I love the changes you're making and how forward thinking you are being. I just want to encourage moving forward one step at a time.

Does it sound good?

XScorpion2 commented:
Ya that's fine, no worries to me as it's just code. I'll make the changes pushing session / group creation down into Audio Device after dinner.

XScorpion2 committed:
Moved session grouping and management into AudioDevice

XScorpion2 commented:
Moved the AudioSession management down into AudioDevice.

t3knomanzer reviewed (commented):
Fantastic! I'm very excited about this, it's looking great.

A few minor things. I think the system sounds logic should be moved to the MainViewModel. The view models are responsible for the data transformation between the services and the views (in our case the device). This will also eliminate the duplicate code in InitializeSystemSessions and InitializeSessions.

The device session is never being sent now. I think you are missing the call to RaiseSessionCreated in the service Initialize when the device is created.

And last, how do you feel about always returning "Master" as the device DisplayName?
Should be easier to identify by the users when navigating in the device. It won't work if we have multiple devices but we can change it when we get there.  
Thoughts?

XScorpion2 commented:
- DisplayName back to "Master" ✔️ 
- RaiseSessionCreated(device) ✔️ (good catch, broke that last min in my refactoring)
- MainViewModel handling of system sounds, not following how that would work cleanly. When you change that value, you still need to go through and find the system sessions and either register or unregister them. Which needs to be done at the AudioDevice level as that's the current owner of that knowledge. There could be data that is stored and tracked in MainViewModel about what is a system session and send it to the device, or remove it from the device from there, however at that point you are duplicating work either way as the AudioDevice and AudioSessionService already cleanly handle session creation and removal at a fairly generic level.

XScorpion2 committed:
Device registration fix & display name change

XScorpion2 commented:
If the concern is about duplicated code between InitializeSessions and InitializeSystemSessions, an idea did come up about how I can reduce that to a single function more easily.

XScorpion2 committed:
Simplified session initialization / reinitialization

t3knomanzer commented:
What I mean is that the device would not have any logic for ignoring the system sounds session, it would always report it.
The MainViewModel would just check in OnAudioSessionCreated if the session created is a systems sound session and report/ignore depending on the setting. At the end of the day, we are just deciding if the session does or doesn't appear in the user interface which is a concern of the ViewModel.  
The MainViewModel needs to know if it's a system sounds session, which can be done in a few different ways, either by exposing the underlying AudioSession property IsSystemsSound or by checking for process id == 0 I believe.

XScorpion2 commented:
Right, but the AudioDevice still needs to re-raise that session created event when the user flips the flag from false, to true. In addition, the wya the ViewModel is set up, it would need to check the IsSystemSound flag on Created, and Volume Changed events, as both get sent to the display and cause them it to wake up. So there still needs to be quite a bit of logic in AudioDevice around handling the flipping of the system sound option or not.

t3knomanzer reviewed (approved):
You're right. Let's leave it as is then!

t3knomanzer commented:
Ready to merge then?

XScorpion2 commented:
I think it's safe to merge if you are good with it's current state. If you really want ViewModel to handle the culling of the system sounds, I'll add it.

t3knomanzer commented:
Nah, I think it looks great already, thank you so much for the fantastic job!

tocklime commented:
I've just compiled and run the latest code on the master branch to test this for the bugs mentioned above. When I open the system volume mixer, and control the levels of an application with the maxmix, I don't see the volume bar go up and down in the system volume mixer, and the sound does not change volume. When I change the volume in system volume mixer, the volume reported on the maxmix /does/ go up and down, and the sound does change volume.

tocklime commented:
...after some more investigation debugging in VS, the `_devices` value in `AudioSessionService` only ever has one entry, corresponding to the master volume. I can't see any code (anywhere) that adds to _devices, except in the initialisation code which adds the master session. Surely `OnSessionCreated` should put something there?

XScorpion2 commented:
@tocklime The management of the various audio sessions are hierarchical based on feedback during this PR. So `_devices` in `AudioSessionService` will only contain `AudioDevice`, so your master volume for your sound card, hdmi display, etc. From there, the `AudioDevice` itself has a `_sessions` field that stores the `AudioSession` and `AudioSessionGroup` that are currently using that `AudioDevice`.

Based on what you have said you are seeing: Controlling volume in Windows makes display volume go up and down, but controlling volume on the display will not make the Windows volume go up and down. I know exactly where I introduced this bug. Whopse

XScorpion2 commented:
@tocklime fixed here: https://github.com/t3knomanzer/maxmix-software/pull/84
